### PR TITLE
DM-21013: Add description for daf_butler expression parser

### PR DESCRIPTION
**Pull request by andy-slac**

### CODE REVIEW

File: doc/lsst.daf.butler/exprParser.rst [hunk 1]
```diff
@@ -0,0 +1,176 @@
+.. _daf_butler_expr_parser:
+
+Butler expression language
+==========================
+
+Butler registry supports user-supplied expression for constraining input,
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: " supports *a* user-supplied"

File: doc/lsst.daf.butler/exprParser.rst [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+output, or intermediate datasets that can appear in the generated
+QuantumGraph. This page describes the structure and syntax of the expression
+language.
+
+Language grammar is defined in ``exprParser.parserYacc`` module which is
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "The language grammar ... module, which is"

File: doc/lsst.daf.butler/exprParser.rst [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+responsible for transforming string with user expression into a syntax tree
+with nodes represented by various classes defined in ``exprParser.exprTree``
+module. Modules in ``exprParser`` package are considered butler/registry
+implementation detail and are not exposed at the butler package level.
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "detail" -> "details"

File: doc/lsst.daf.butler/exprParser.rst [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+In the initial implementation of the registry the expression was pasted into
+a giant SQL query that selected datasets and dimensions from registry database.
+Before pasting it into a query the expression was parsed and thus checked for
+possible errors. The language for expression was at that point decided to be
+a small subset of the expressions allowed in SQL WHERE clause. Later we
+started generating SQL code from a parsed syntax tree and extended the
+language with few additional constructs that are not in SQL.
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: I'm not sure the history is relevant for a user guide; I think I'd just say that the language is similar to a subset of SQL with some extensions.

File: doc/lsst.daf.butler/exprParser.rst [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+Expression structure
+--------------------
+
+The expression is passed as a string to a registry methods that build a
+QuantumGraph typically from a command-line application such as ``pipetask``.
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst:25]: It probably doesn't make sense to change this yet, as I imagine this ticket will land before DM-17023, but when DM-17023 does land, we should remember to update this paragraph (and the very first one) to reflect the fact that the expression language is used in more than just QuantumGraph generation there.

File: doc/lsst.daf.butler/exprParser.rst [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+The string contains a single boolean expression which evaluates to true or
+false (if it is a valid expression). Expression can contain a bunch of
+standard logical operators, comparisons, literals, and identifiers which are
+references to registry objects.
+
+Few words in expression grammar are reserved, these are ``AND``, ``OR``,
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "A few words...are reserved: AND, OR"

File: doc/lsst.daf.butler/exprParser.rst [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+``NOT`` and ``IN``. Reserved words are not case sensitive and can appear in
+either upper or lower case, or a mixture of both.
+
+Language operator precedence rules are the same as for the other languages
+like C++ or Python. When in doubt use grouping operators (parentheses) for
+sub-expressions.
+
+General note - parser itself does not evaluate any expressions even if they
+consist of literals only, all evaluation happens in SQL engine when registry
+runs resulting SQL query.
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "the parser"..."the SQL engine"..."the resulting SQL query"

File: doc/lsst.daf.butler/exprParser.rst [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+Following sections describe each of the parts in detail.
+
+Literals
+--------
+
+The language supports these types of literals:
+
+Strings
+    This is just a sequence of characters enclosed in single quotation marks.
+    The parser itself fully supports Unicode but some tools such as database
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "Unicode, but"..."for it, depending"

File: doc/lsst.daf.butler/exprParser.rst [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+Numbers
+    Integer numbers are series of decimal numbers optionally preceded by
+    minus sign. Parser does not support octal/hexadecimal numbers. Floating
+    point numbers use standard notation with decimal point and/or exponent.
+    For numbers parser passes a string representation of a number to
+    downstream registry code to avoid possible rounding issues.
+
+Range literals
+    This sort of literals is allowed inside ``IN`` expression only. It consists
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "literal is"..."expressions only"

File: doc/lsst.daf.butler/exprParser.rst [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    of two integer literals separated by double dots and optionally followed by
+    a colon and one more integer literal. Two integers define start and stop
+    values for the range, both are inclusive values. Optional third integer
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "range; both"..."The optional third"

File: doc/lsst.daf.butler/exprParser.rst [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    defines stride value which defaults to 1, it cannot be negative. Ranges
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "value, which"..."to 1; it cannot"

File: doc/lsst.daf.butler/exprParser.rst [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+Examples of range literals:
+
+* ``1..5`` -- equivalent to ``1,2,3,4,5``
+* ``1..10:3`` -- equivalent to ``1,4,7,10``
+* ``-10..-1:2`` -- equivalent to ``-10,-8,-6,-4,-2``
+
+Identifiers
+-----------
+
+Identifiers represent values external to a parser such as values stored in a
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "parser, such as"

File: doc/lsst.daf.butler/exprParser.rst [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+database. Parser itself cannot define identifiers or their values, it is
+responsibility of translation layer (registry) to map identifiers into
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "The parser itself"..."values; it is the responsibility of the translation later"

File: doc/lsst.daf.butler/exprParser.rst [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+something sensible. Like in most programming languages identifier starts with
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "languages, an identifier"

File: doc/lsst.daf.butler/exprParser.rst [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+The ``IN`` operator (and ``NOT IN``) are an expanded version of a regular SQL
+IN operator. Its general syntax looks like::
+
+    <expression> IN ( <literal1>[, <literal2>, ... ])
+    <expression> NOT IN ( <literal1>[, <literal2>, ... ])
+
+where each item in the right hand side list is one of the supported literals.
+Unlike regular SQL IN operator the list cannot contain expressions, only
+literals. The extension to regular SQL IN is that literals can be range
+literals as defined above. And it can also be a mixture of integer literals
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "above. It can also be"

File: doc/lsst.daf.butler/exprParser.rst [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+and range literals (language allows mixing of string literals and ranges but
+it may not make sense when translated to SQL).
+
+For an example of range uses these two expressions are equivalent::
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst]: "range usage, these two"

File: doc/lsst.daf.butler/exprParser.rst [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+``*`` (multiply), ``/`` (divide), and ``%`` (modulo). Usual precedence rules
+apply to these operators. Operands for them can be anything that evaluates to
+a numeric value.
+
+Comparison operators
+--------------------
+
+Language supports set of regular comparison operators - ``=``, ``!=``, ``<``,
+``<=``, ``>``, ``>=``. This can be used on operands that evaluate to a numeric
+values, for (in)equality operators operands can also be boolean expressions.
```
Comment (TallJimbo) [inline doc/lsst.daf.butler/exprParser.rst:117]: Maybe add a note that equality comparison is a single `=`, like SQL but unlike Python?

**Review by TallJimbo — APPROVED** [review]
Looks good!  Most comments are just minor grammar/typos issues.

File: doc/lsst.daf.butler/exprParser.rst (hunk 17 above)
Comment (andy-slac) [reply doc/lsst.daf.butler/exprParser.rst:117]: I added this:

    .. note :: The equality comparison operator is a single ``=`` like in SQL, not
        double ``==`` like in Python or C++.

**Pull request merged**